### PR TITLE
Software architecture

### PR DESCRIPTION
**Pull request by jschlenger**
Adding a new part to the OIP documentation describing the overall software architecture and the express binding generator.

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ It comes with its own `EXPRESS` parser and `C++` classes generator.
 - [CMake Options Documentation](Documentation/markdown/CMakeOptions.md)
 - [Git process Guideline](Documentation/markdown/GitProcess.md)
 - [Coding Guideline](Documentation/markdown/CodingGuideline.md)
+- [Software Architecture](Documentation/markdown/SoftwareArchitecture.md)
```
Comment (pjanck) [inline Readme.md]: I'd put it as the second entry in this list

**Review by pjanck — CHANGES_REQUESTED** [review]
Please amend as discussed offline.

File: Readme.md (hunk 1 above)
Comment (jschlenger) [reply Readme.md]: Resolved: 30a64e9

File: Documentation/markdown/SoftwareArchitecture.md [hunk 2]
```diff
@@ -15,63 +16,119 @@
...
-* Carve (constructive solid geometry library)
-* Tclap (small and flexible library providing an interface for accessing and defining command line arguments; header-only)
-* FFTW3 (computation of Fourier transformations; header-only)
-* Doxygen (tool to generate documentation; header-only)
+The base module constitutes the backbone of the Open Infra Platform, 
+with all of the other software parts building upon it. 
+As its main part, several third-party libraries are linked here. 
+Other than that, it also provides very fundamental elements like the _IModel_ interface 
+that is implemented by the upper layers for every supported file format (e.g., `IFC` and `OFF`)
+and holds all of the information read from the input files.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]:
```suggestion
and provides some basic information about the input files. 
```

It doesn't hold all the information. As a matter of fact, it holds no information. It only *promises*  that some information will be made availlable by the implementation.

File: Documentation/markdown/SoftwareArchitecture.md [hunk 3]
```diff
@@ -15,63 +16,119 @@
...
-The base module constitutes the backbone of the Open Infra Platform, with all of the other software parts building upon it. As its main part, several third-party libraries are linked here. Other than that, it also provides very fundamental elements like the IModel interface that is implemented by the upper layers for every supported file format (e.g., IFC and OFF) and will hold all of the information read from the input files. Furthermore, classes and structures for georeferencing metadata and simple bounding boxes are given.
-
-Linked third-party libraries:
-* Boost (library for linear algebra, multithreading, and storage management; header-only)
-* BlueFramework (open-source C++ 3D graphics rendering engine developed by TUM)
-* Carve (constructive solid geometry library)
-* Tclap (small and flexible library providing an interface for accessing and defining command line arguments; header-only)
-* FFTW3 (computation of Fourier transformations; header-only)
-* Doxygen (tool to generate documentation; header-only)
+The base module constitutes the backbone of the Open Infra Platform,
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]:
```suggestion
The _Base_ module constitutes the backbone of the Open Infra Platform, 
```

Consistency :)

File: Documentation/markdown/SoftwareArchitecture.md [hunk 4]
```diff
@@ -15,63 +16,119 @@
...
-* CloudCompare (point cloud processing)
-
-By default, the Point Cloud Processing module is deactivated. In order to make use of it, the OIP needs to be built with the corresponding CMake option selected (for further details, see [CMake Options Documentation](Documentation/markdown/CMakeOptions.md)). A significant advantage is that the Core and the Point Cloud Processing module can be run simultaneously so that, e.g., one can open an IFC file and a point cloud at the same time to compare them to each other directly.
+The modules can be separated into two central parts. 
+First of all, the _Core_ module links to the third-party libraries `Eigen` and `Qt`.
+It also comprises one submodule for the IFC (Industry Foundation Classes) and OFF (Object File Format) file format each. 
+In the case of IFC, the _IFC Geometry Converter_ takes care of the geometric conversion processes. 
+It makes use of the _Express Binding Library_, which is generated by the [Express Binding Generator](#Express_binding). 
+The IFC file format supports all sorts of geometric representations
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md:44]:
```suggestion

The IFC file format supports all sorts of geometric representations 
```

I'd start a new paragraph here.

File: Documentation/markdown/SoftwareArchitecture.md [hunk 5]
```diff
@@ -15,63 +16,119 @@
...
+the _IFC Geometry Converter_ has to convert the geometric descriptions 
+specified in IFC into a triangle mesh that can be understood by the rendering engine.
+This step is based on the geometry kernel `carve`. 
+Since IFC encompasses a multitude of representations and is constantly updating, 
+only a subset of the _IfcRepresentationItems_ are supported at the moment, 
+focusing on the most frequently used ones. 
+However, OIP is continuously expanding to cover more and more types of representations. 
+To check which representation items are currently supported, 
+the user can take a look at the [List of supported IFC Representations](./SupportedIFCrepresentations.md). 
+The other part of the _Core_ module consists of the _OFF Geometry Converter_.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md:58]:
```suggestion

The other part of the _Core_ module consists of the _OFF Geometry Converter_. 
```

File: Documentation/markdown/SoftwareArchitecture.md [hunk 6]
```diff
@@ -15,63 +16,119 @@
...
+so that, e.g., one can open an IFC file and a point cloud 
+at the same time to compare them to each other directly.
 
 ### <a name="UI"></a> User Interface
 
-The underlying part of the user interface is the Rendering module. This contains a self-written rendering engine called BlueFramework. Depending on the objects that are to be displayed, different types of shaders and effects are implemented. The Graphical User Interface is built on top of this, which has a 3D viewport to display geometric objects and several menus and buttons that provide functions to interact with them. The GUI bases on the Qt library, which has an extensive catalog of ready-to-use user interface elements. Other third-party libraries that the GUI uses are:
-* Zlib (algorithms for compression and decompression of data)
-* Proj (functionalities to calculate conversions between geodetic reference systems)
+The underlying part of the user interface is the _Rendering module_. 
+This links to a self-written rendering engine called _BlueFramework_.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]: Link to the repo would be best here.

File: Documentation/markdown/SoftwareArchitecture.md [hunk 7]
```diff
@@ -15,63 +16,119 @@
...
 
 ### <a name="UI"></a> User Interface
 
-The underlying part of the user interface is the Rendering module. This contains a self-written rendering engine called BlueFramework. Depending on the objects that are to be displayed, different types of shaders and effects are implemented. The Graphical User Interface is built on top of this, which has a 3D viewport to display geometric objects and several menus and buttons that provide functions to interact with them. The GUI bases on the Qt library, which has an extensive catalog of ready-to-use user interface elements. Other third-party libraries that the GUI uses are:
-* Zlib (algorithms for compression and decompression of data)
-* Proj (functionalities to calculate conversions between geodetic reference systems)
+The underlying part of the user interface is the _Rendering module_. 
+This links to a self-written rendering engine called _BlueFramework_. 
+Depending on the objects that are to be displayed, different types of shaders and effects are implemented. 
+The _Graphical User Interface_ is built on top of this,
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md:83]:
```suggestion

The _Graphical User Interface_ is built on top of this, 
```

File: Documentation/markdown/SoftwareArchitecture.md [hunk 8]
```diff
@@ -15,63 +16,119 @@
...
+The GUI bases on the `Qt` library, which has an extensive catalog of ready-to-use user interface elements. 
+Other third-party libraries that the GUI uses are Zlib and Proj (for more details see [here](Licenses/readme.md))
 
-On the other hand, there is the Unit Testing part, which bases on Google Test. Since many unit tests that are implemented in the OIP use visual comparison, the unit test base includes a geometry model renderer that allows rendering without the use of the Graphical User Interface. On top of that, there are various individual unit tests for IFC, OFF, and point clouds (and also some other more general unit tests?). These provide automated checking to guarantee that implemented functionalities are all still in place and working as expected. For the case of the IFC format, the [List of supported IFC Representations](./SupportedIFCrepresentations.md) provides information about which representations currently have at least one unit test in place.
+### <a name="UT"></a> Unit Testing
+
+On the other hand, there is the _Unit Testing_ part, which bases on `Google Test`.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]: Careful - if one uses *on the other hand*  there **must** be *on the one hand* somewhere as well.

File: Documentation/markdown/SoftwareArchitecture.md [hunk 9]
```diff
@@ -15,63 +16,119 @@
...
 ## <a name="Express_binding"></a> Express Binding Generator
 
-This section will give a short overview of how the Express Binding Generator is working. The general task of the Express Binding Generator is to read an IFC schema in the form of an EXPRESS file and automatically create C++ classes mirroring the IFC entities. This allows to adapt and support the newest IFC versions quickly. The Express Binding Generator consists of three main parts: the executables, the commands, and the early binding library. These will be explained in more detail in the following.
+This section gives a short overview of how the _Express Binding Generator_ is working. 
+The general task of the _Express Binding Generator_ is to read an IFC schema in the form of an EXPRESS schema definition 
+and automatically create early binding C++ classes mirroring the IFC entities. 
+This allows to find coding errors already at compile time. 
+The _Express Binding Generator_ consists of three main parts: 
+the executables, the commands, and the early binding library. 
+These are explained in more detail in the following.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]: in the following *what*?

File: Documentation/markdown/SoftwareArchitecture.md [hunk 10]
```diff
@@ -15,63 +16,119 @@
...
-The executables make use of the following third-party libraries:
-* Flex (lexical analyzer generator)
-* Bison (general-purpose parser generator)
-* Tclap (small and flexible library providing an interface for accessing and defining command line arguments; header-only)
-* Boost (library for linear algebra, multithreading, and storage management; header-only)
+![](../images/express_parser.png)
+
+In the executables, the basic functionalities are defined, which state how the IFC EXPRESS schemas are converted to C++ classes,
+taking into account entities, inverse attributes, attributes, types, and SELECT clauses. 
+A Flex and Bison parser generates a corresponding meta-model representation in C++,
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md]:
```suggestion
Flex and Bison generate a corresponding meta-model representation in C++, 
```

File: Documentation/markdown/SoftwareArchitecture.md [hunk 11]
```diff
@@ -15,63 +16,119 @@
...
-* Tclap (small and flexible library providing an interface for accessing and defining command line arguments; header-only)
-* Boost (library for linear algebra, multithreading, and storage management; header-only)
+![](../images/express_parser.png)
+
+In the executables, the basic functionalities are defined, which state how the IFC EXPRESS schemas are converted to C++ classes,
+taking into account entities, inverse attributes, attributes, types, and SELECT clauses. 
+A Flex and Bison parser generates a corresponding meta-model representation in C++, 
+which needs to overcome fundamental differences between C++ and EXPRESS.
+The executables make use of the third-party libraries Flex, Bison, Tclap, and Boost ([more details](Licenses/readme.md)))
+Not all parts of the EXPRESS schema are supported at the moment.
```
Comment (pjanck) [inline Documentation/markdown/SoftwareArchitecture.md:120]:
```suggestion

Not all parts of the EXPRESS schema are supported at the moment.
```

**Review by pjanck — APPROVED** [review]
Some suggestions, otherwise very well made! When corrected, I believe that we can merge this.

File: Documentation/markdown/SoftwareArchitecture.md (hunk 2 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 3 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 4 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md:44]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 6 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 7 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md:83]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 9 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 10 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 11 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md:120]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 8 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md]: Resolved: 7aaecb6

File: Documentation/markdown/SoftwareArchitecture.md (hunk 5 above)
Comment (jschlenger) [reply Documentation/markdown/SoftwareArchitecture.md:58]: Resolved: 7aaecb6

**Pull request merged**